KAFKA-6782: solved the bug of restoration of aborted messages for GlobalStateStore and KGlobalTable

Gitomain opened the pull request.

Gitomain committed:
resolve the bug of KAFKA-6782, the restoration of aborted messages for GlobalStateStore

mjsax commented:
\cc @bbejeck @vvcephei

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:
```diff
@@ -261,16 +265,23 @@ public boolean conditionMet() {
         CLUSTER.createTopic(globalTableTopic, 2, 1);
     }
-
+    
```
Please avoid unnecessary reformatting -- there is more in this PR -- guess, it's a matter of IDE settings?

Gitomain replied:
Yes, it's a matter of IDE settings, we don't need these chages.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:
```diff
@@ -110,6 +112,7 @@ public void before() throws InterruptedException {
         streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
         streamsConfiguration.put(IntegrationTestUtils.INTERNAL_LEAVE_GROUP_ON_CLOSE, true);
         streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, "exactly_once");
```
I think it would be better, to have two test -- one with EOS enabled and one without.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:
```diff
@@ -261,16 +265,23 @@ public boolean conditionMet() {
+    
     private void startStreams() {
-        kafkaStreams = new KafkaStreams(builder.build(), streamsConfiguration);
+        Topology toplogy = builder.build();
```
Why this change? We don't need the variable `topology`?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:
```diff
@@ -291,6 +302,25 @@ private void produceTopicValues(final String topic) throws Exception {
                 mockTime);
     }
 
+    private void produceAbortedMessages() throws ExecutionException, InterruptedException {
```
nit: in tests, it better to just declare `throws Exception` -- it does not provide value to spell out the different types in test and introduces "noise"

I realized, that the existing code also does declare different types of exception. Would you mind to clean this up, too?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:
```diff
@@ -291,6 +302,25 @@ private void produceTopicValues(final String topic) throws Exception {
     }
 
+    private void produceAbortedMessages() throws ExecutionException, InterruptedException {
+        Properties properties = new Properties();
```
nit: add `final` (we try to use `final` whenever possible)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -140,16 +140,33 @@ public static void purgeLocalStreamsState(final Properties streamsConfiguration)
     }
+    
+    public static <K, V> void produceAbortedKeyValuesSynchronouslyWithTimestamp(final String topic,
+            final Collection<KeyValue<K, V>> records,
```
nit: formatting -- parameters should be aligned

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -140,16 +140,33 @@ public static void purgeLocalStreamsState(final Properties streamsConfiguration)
+            final Collection<KeyValue<K, V>> records,
+            final Properties producerConfig,
+            final Long timestamp) throws ExecutionException, InterruptedException {
+        try (Producer<K, V> producer = new KafkaProducer<>(producerConfig)) {
```
nit: add `final`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -140,16 +140,33 @@ public static void purgeLocalStreamsState(final Properties streamsConfiguration)
+    public static <K, V> void produceAbortedKeyValuesSynchronouslyWithTimestamp(final String topic,
+            final Collection<KeyValue<K, V>> records,
+            final Properties producerConfig,
+            final Long timestamp) throws ExecutionException, InterruptedException {
```
nit: exceptions

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:
```diff
@@ -261,16 +265,23 @@ public boolean conditionMet() {
     }
+    
+    @Test(timeout = 10000)
+    public void shouldRestoreAbortedMessages() throws ExecutionException, InterruptedException {
```
nit: the name is a little miss leading -- aborted message should not be restored...

It also seems, this test only checks if the restore terminates -- I am wondering, if we should also write committed data into the topic and check that the messages got restored? Not sure if necessary.

mjsax reviewed (commented):
Thanks for the PR. Couple of comments.

Side note: do we need all those `.gitignore` changes?

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -159,12 +176,11 @@ public static void purgeLocalStreamsState(final Properties streamsConfiguration)
-                                                                                  final String topic,
-                                                                                  final int expectedNumRecords) throws InterruptedException {
-
+                                                                                      final String topic,
```
nit: formatting - align parameters

mjsax replied:
please address this comment, too

bbejeck reviewed (commented):
Overall looks good just left a minor comment.  I agree with @mjsax about the `.gitignore` changes, can we double check those?

Gitomain committed:
Update .gitignore

Gitomain committed:
Update .gitignore

mjsax commented:
@Gitomain what is the status of this PR -- seems there are still couple of comments you did not address yet.

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
Delete .gitignore

Gitomain committed:
added integration test GlobalKTableEOSIntegrationTest to test the EOS enabled case

Gitomain committed:
merge the delete of .gitignore files
Merge branch 'aborted_messages_restoring' of https://github.com/Gitomain/kafka into aborted_messages_restoring

Gitomain commented:
Hello, I have deleted all the .gitignore files and added a new test GlobalKTableEOSIntegrationTest to test the EOS case independently.

guozhangwang commented:
There is a recent commit that moves which package `Consumed` belongs, causing the jenkins to fail:

```
17:02:22 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:26: error: cannot find symbol
17:02:22 import org.apache.kafka.streams.Consumed;
17:02:22                                ^
17:02:22   symbol:   class Consumed
17:02:22   location: package org.apache.kafka.streams
17:02:25 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:114: error: cannot find symbol
17:02:25         globalTable = builder.globalTable(globalTableTopic, Consumed.with(Serdes.Long(), Serdes.String()),
17:02:25                                                             ^
17:02:25   symbol:   variable Consumed
17:02:25   location: class GlobalKTableEOSIntegrationTest
17:02:25 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:118: error: cannot find symbol
17:02:25         final Consumed<String, Long> stringLongConsumed = Consumed.with(Serdes.String(), Serdes.Long());
17:02:25               ^
17:02:25   symbol:   class Consumed
17:02:25   location: class GlobalKTableEOSIntegrationTest
17:02:25 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:118: error: cannot find symbol
17:02:25         final Consumed<String, Long> stringLongConsumed = Consumed.with(Serdes.String(), Serdes.Long());
```

Could you rebase your branch on latest trunk head and merge any conflicts?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -140,16 +140,34 @@ public static void purgeLocalStreamsState(final Properties streamsConfiguration)
     public static <V> void produceValuesSynchronously(
-        final String topic, final Collection<V> records, final Properties producerConfig, final Time time)
-        throws ExecutionException, InterruptedException {
+            final String topic, final Collection<V> records, final Properties producerConfig, final Time time)
```
Are those changes intended? Or are they IDE enforced?

Gitomain replied:
I think it's a problem of IDE

guozhangwang reviewed (commented):
This is a nice catch, thanks @Gitomain . One meta question: what's the motivation of extracting the EOS test into a separate class?

Gitomain commented:
For the motivation of extracting the EOS test into a separate class, I think both the EOS enabled and the EOS disabled cases should be tested. And I found it was a little messy to test the 2 cases in the same class. So I extracted the EOS test into a separate class.

Gitomain committed:
Merge remote-tracking branch 'upstream/trunk' into aborted_messages_restoring

Gitomain committed:
update to kafka trunk

Gitomain committed:
change .gitignore

Gitomain committed:
Merge branch 'aborted_messages_restoring' of https://github.com/Gitomain/kafka into aborted_messages_restoring

Gitomain committed:
deleted all the .class files

Gitomain committed:
corrected comsumed Consumed import problem

bbejeck commented:
@Gitomain why include `log4j` files for connect and tools?

Gitomain committed:
Delete log4j.properties

Gitomain committed:
Delete log4j.properties

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+        final String applicationId = "globalTableTopic-table-eos-test-" + testNo;
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration
+                .put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
```
nit: formatting

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, "exactly_once");
+        globalTable = builder.globalTable(globalTableTopic, Consumed.with(Serdes.Long(), Serdes.String()),
+                                          Materialized.<Long, String, KeyValueStore<Bytes, byte[]>>as(globalStore)
```
with Java 8 support, I think we can remove generics here

Gitomain replied:
Hello, I can't understand how can we remove generics here, so I didn't modify this part.

mjsax replied:
Java8 should be able to infer the type and Java7 was dropped recently. So we want to update the code to Java8 incrementally. You should be able to change this to:
```
Materialized.as(globalStore)
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -179,16 +179,34 @@ public static void purgeLocalStreamsState(final Properties streamsConfiguration)
+                f.get();
+                producer.abortTransaction();
+            }
+            producer.flush();
```
do we need a `flush()` here? `abortTransaction()` is a sync call

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+                return result.equals(expected);
+            }
+        }, 30000L, "waiting for initial values");
+        System.out.println("no failed test");
```
this line can be removed

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+        System.out.println("no failed test");
+    }
+    
+    @Test(timeout = 40000)
```
do we need the timeout? `waitForCondition` already applies a timeout

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+                mockTime.milliseconds());
+    }
+
+    private void produceInitialGlobalTableValues() throws Exception {
```
this method should enable EOS by default for this class, shouldn't it?

Gitomain replied:
Thanks for remind, it's done

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+        produceInitialGlobalTableValues(false);
+    }
+
+    private void produceInitialGlobalTableValues(final boolean enableTransactions) throws Exception {
```
we can merge this method into the one above as we always want EOS enabled?

mjsax reviewed (commented):
Thanks for adding the test!

I think it's ok to split both tests. However, we might want to extract shared code into an abstract super class for both.

Some more follow up comment.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+
+    private static volatile int testNo = 0;
+    private final MockTime mockTime = CLUSTER.time;
+    private final KeyValueMapper<String, Long, Long> keyMapper = new KeyValueMapper<String, Long, Long>() {
```
Just one meta-comment, now that Streams supports Java 8 do we want to covert the use of anonymous class usage here and other places to lambda expressions?  This comment is very opinionated, and this PR has been going on for a while now, so I wouldn't hold up merging for this.

Gitomain replied:
Hello, I know what you mean. But I have seen that all the test codes are still in this style now, I prefer to leave it as it is, what do you think ?

mjsax replied:
As mentioned in the other comment. Java7 was just dropped recently and thus most code is still Java7. We want to incrementally rewrite to Java8. Thus, feel free to update the code accordingly.

bbejeck reviewed (commented):
Just left one last meta-comment.  Overall lgtm, pending addressing @mjsax comments

mjsax commented:
@Gitomain Code freeze for 2.0 release is on Tuesday (6/12). Can you finish this PR so we can get it into the release? (If not, I might just take it and apply the last changes myself to merged it.) Please let us know.

Gitomain committed:
done some corrections about the format

Gitomain committed:
Merge branch 'aborted_messages_restoring' of https://github.com/Gitomain/kafka into aborted_messages_restoring

Gitomain committed:
additional corrections

Gitomain commented:
Hello @mjsax , I have done some modification. But I have no time to extract shared code into an abstract class before 6/12. I think we can finish this PR now.

mjsax commented:
The deadline was pushed out by one day. I think we can merge as-is. The code rewrites to Java8 are not too important.

mjsax commented on .gitignore:
```diff
@@ -1,5 +1,5 @@
 dist
-*classes
+*.class
```
@Gitomain  We can add `*.class` but we need to keep `classes` as some folders are named like this. Can you update the PR? Otherwise I can fix during merging.

Gitomain committed:
added '*classes' in .gitignore

bbejeck reviewed (approved):
Thanks for the patch, LGTM

mjsax commented:
Thanks for the fix @Gitomain 

Merged to `trunk` and cherry-picked to `2.0`, `1.1`, `1.0`, and `0.11.0`.
